server/WAPI.rb

```diff
@@ -0,0 +1,189 @@
+        if j.has_key?('responseCode')
+          j['responseCode'] = j['responseCode'].to_i
+        end
+      rescue => err
+        logger.info "WAPI: #{__LINE__}: do_request: conversion error "+j.inspect
+        ## because of the error reset back to the original response.
+        j = JSON.parse(response)
+      end
+      ## get response code from nested response or from the original response
+      rc = j['responseCode'] || response.code
+      j = JSON.generate(j)
+      wrapped_response = WAPIResultWrapper.new(rc, "COMPLETED", j)
+    rescue Exception => exp
+      logger.debug "WAPI: #{__LINE__}: do_request: exception: "+exp.inspect
+      wrapped_response = WAPIResultWrapper.new(666, "EXCEPTION", exp)
+    end
+    #logger.debug "WAPI: #{__LINE__}: do_request: wrapped response: "+wrapped_response.inspect
+    r.stop
+    logger.info "WAPI: #{__LINE__}: do_request: stopwatch: "+ r.pretty_summary
+    wrapped_response
+  end
+
+  ## Run the request.  If the result is unauthorized then renew the token and try again.
+  ## In any case will return a wrapped result.
+
+  def get_request(request)
+    wrapped_response = do_request(request)
+
+    ## If appropriate try to renew the token.
+    if wrapped_response.meta_status == 666 &&
+        wrapped_response.result.respond_to?('http_code') &&
+        wrapped_response.result.http_code == 401
+      #logger.debug("WAPI: #{__LINE__}: unauthorized on initial request: "+wrapped_response.inspect)
+      wrapped_response = renew_token()
+      ## if the token renewed ok then try the request again.
+      if wrapped_response.meta_status == 200
+#        logger.debug("WAPI: #{__LINE__}: retrying request after token renewal")
+        wrapped_response = do_request(request)
+      end
+    end
+    wrapped_response
+  end
+
+  # Renew the current token.  Will set the current @token value in the object
+  def renew_token
+
+    #logger.info "WAPI: renew_token"
+    begin
+      renew = Stopwatch.new("renew token")
+      renew.start;
+      response = RestClient.post @token_server,
+                                 "grant_type=client_credentials&scope=PRODUCTION",
+                                 {
+                                     :Authorization => @renewal,
+                                     :content_type => "application/x-www-form-urlencoded"
+                                 }
+      ## If it worked then parse the result.  This is here to capture any JSON parsing exceptions.
+      if response.code == 200
+        ## will need to get the access_token below.  If it is not JSON that is an error.
+        s = JSON.parse(response)
+        @token = s['access_token']
+      end
+    rescue Exception => exp
+      # If got an exception for the renewal then wrap that up to be returned.
+      logger.info("WAPI: #{__LINE__}: renewal post exception: "+exp.to_json+":"+exp.http_code.to_s)
+      wr = WAPIResultWrapper.new(exp.http_code, "EXCEPTION DURING TOKEN RENEWAL", exp)
+      renew.stop
```

Comment:
same as above comment, move the stop line to right after the RestClient.post call.